### PR TITLE
Adding logic to initialize tclArgs on settings load

### PR DESCRIPTION
**Pull request by skyler-rs**
> ### Motivate of the pull request
> Adding missing feature in settings
>
> ### Describe the technical details
> #### What is currently done? (Provide issue link if applicable)
> Prior to this change, widget factory settings saved in a previous session wouldn't be set on a program reload until the user opened and applied/saved the settings.
>
> #### What does this pull request change?
> - On settings load, the json is now traversed to find any tcl arg values and those values are set using their associated tclSetter using either default or stored userValues.
> - This also adds unit tests for the settings system

> ### Which part of the code base require a change
> 
> - [ ] Library: <Specify the library name>
> - [ ] Plug-in: <Specify the plugin name>
> - [ ] Engine
> - [ ] Documentation
> - [x] Regression tests
> - [ ] Continous Integration (CI) scripts

> ### Impact of the pull request
> Tcl settings will be initialized at start

> ### Testing
> - Unit tests have been added for the settings system
> - Raptor integration was manually tested

### CODE REVIEW

File: src/Main/Settings.cpp [hunk 1]
```diff
@@ -166,3 +170,141 @@ void Settings::loadJsonFile(json* jsonObject, const QString& filePath) {
...
+  json& topJson = getJson();
+
+  // Step through each settings path and load tcl vars if possible
+  for (QString path : paths) {
+    // Get the json stored at this path
+    json::json_pointer jsonPtr(path.toStdString());
+    json& settingsJson = topJson.at(jsonPtr);
+
+    // Look for a tclArgeKey
+    QString tclArgKey = "";
```
Comment (KochynVolodymyr) [inline src/Main/Settings.cpp:190]: same here

File: src/Main/Settings.cpp [hunk 2]
```diff
@@ -166,3 +170,141 @@ void Settings::loadJsonFile(json* jsonObject, const QString& filePath) {
...
+  traverseJson(jsonData, findCb);
+  return jsonPaths;
+}
+
+// This will traverse a json object looking for an "arg" field, if it finds one
+// it will then check the current json object for "userValue" and "default"
+// fields which it will use to create a tcl arg list string from
+QString Settings::getTclArgString(json& jsonData) {
+  // Create a callback to collect and return tcl arg strings
+  QString argStr = "";
```
Comment (KochynVolodymyr) [inline src/Main/Settings.cpp:245]: same here

File: src/Main/Settings.cpp [hunk 3]
```diff
@@ -166,3 +170,141 @@ void Settings::loadJsonFile(json* jsonObject, const QString& filePath) {
...
+  auto findCb = [&argStr](json& obj, const QString& path) {
+    if (obj.type() == nlohmann::detail::value_t::object) {
+      QString widgetType = QString::fromStdString(obj.value("widgetType", ""));
+
+      // Check if the object has an "arg" field
+      QString tclArg = QString::fromStdString(obj.value("arg", ""));
+      if (!tclArg.isEmpty()) {
+        QString val = "";
```
Comment (KochynVolodymyr) [inline src/Main/Settings.cpp:253]: same here

File: src/Main/Settings.cpp [hunk 4]
```diff
@@ -166,3 +170,141 @@ void Settings::loadJsonFile(json* jsonObject, const QString& filePath) {
...
+
+        // Create a safe json look up that will read the expected value type
+        // based off the widgetType
+        auto getValStr = [obj, widgetType](const std::string& key) -> QString {
+          QString val = "";
```
Comment (KochynVolodymyr) [inline src/Main/Settings.cpp:258]: same here

File: tests/unittest/Settings/Settings_test.cpp [hunk 5]
```diff
@@ -0,0 +1,261 @@
...
+TEST(Settings, loadSettings_merges_json) {
+  Settings test;
+
+  // The goal of this test is to load two json files w/ different child nodes
+  // and ensure that the two files are combined
+
+  // This implicitly tests Settings::loadJsonFile() which is used by
+  // loadSettings() to load json files
+
+  QStringList files{"../tests/unittest/Settings/settings_loadJson_1.json",
```
Comment (KochynVolodymyr) [inline tests/unittest/Settings/Settings_test.cpp:46]: why not just ...{settings_loadJson_1.json, ....} ?

Comment (skyler-rs) [reply tests/unittest/Settings/Settings_test.cpp:46]: I get the following when i remove the relative path info:
ERROR - Settings::loadJsonFile - Failed to read "settings_loadJson_1.json"
ERROR - Settings::loadJsonFile - Failed to read "settings_loadJson_2.json"

File: tests/unittest/Settings/Settings_test.cpp [hunk 6]
```diff
@@ -0,0 +1,261 @@
...
+#include "Main/WidgetFactory.h"
+#include "gmock/gmock.h"
+#include "gtest/gtest.h"
+
+using ::testing::ElementsAre;
+
+namespace FOEDAG {
+namespace {
+
+TEST(Settings, loadSettings_merges_json) {
```
Comment (ravic-rs) [inline tests/unittest/Settings/Settings_test.cpp:37]: Please update the test case name as per gtest's recommendation.
https://google.github.io/googletest/primer.html#simple-tests

**Review by ravic-rs — APPROVED** [review]
Refactor in future to incorporate setup. teardown and test_f features.

File: tests/unittest/Settings/Settings_test.cpp (hunk 5 above)
Comment (KochynVolodymyr) [reply tests/unittest/Settings/Settings_test.cpp:46]: This doesn't work from Qt creator.

**Comment by KochynVolodymyr** [comment]
There is an issue with test files path. It doesn't work locally.

**Comment by ravic-rs** [comment]
On hold till #627  is merged for moving the file paths to resources.qrc: Supporting gtests using QtCreator.

**Comment by ravic-rs** [comment]
> On hold till #627 is merged for moving the file paths to resources.qrc: Supporting gtests using QtCreator.

@skyler-rs #627 has been merged already. When can we close this ?

**Comment by skyler-rs** [comment]
> 

Just updated the unittests, should be good to check-in after CI's pass

**Comment by ravic-rs** [comment]
> > @skyler-rs #627 has been merged already. When can we close this ?
> 
> Just updated the unittests, should be good to check-in after CI's pass

Did we verify these tests are running on the CI ? One of the recent PRs from Volodymyr, the CI was successful but the test never ran. So I would like to double-check that. This can't be rebased, if we can't fix then I will have to squash the commits.

**Comment by skyler-rs** [comment]
> > > @skyler-rs https://github.com/os-fpga/FOEDAG/pull/627 has been merged already. When can we close this ?

> > Just updated the unittests, should be good to check-in after CI's pass

> Did we verify these tests are running on the CI ? One of the recent PRs from Volodymyr, the CI was successful but the test 
never ran. So I would like to double-check that. This can't be rebased, if we can't fix then I will have to squash the commits.

It looks like it ran according to the logs?
image

**Comment by skyler-rs** [comment]
Re the merge, github has all merge options greyed out for me, could you squash?

**Pull request merged**